Build error when putting a value of a sum type involving a function type in a map

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->
**V doctor:**
```
OS: linux, Pop!_OS 22.04 LTS
Processor: 4 cpus, 64bit, little endian, Intel(R) Core(TM) i7-5600U CPU @ 2.60GHz
CC version: cc (Ubuntu 11.2.0-19ubuntu1) 11.2.0

getwd: /home/hedo/usr/tmp/v
vmodules: /home/hedo/.vmodules
vroot: /home/hedo/usr/tmp/v/v
vexe: /home/hedo/usr/tmp/v/v/v
vexe mtime: 2022-09-06 20:59:53
is vroot writable: true
is vmodules writable: true
V full version: V 0.3.1 4bd49a0.ca99a1d

Git version: git version 2.34.1
Git vroot status: weekly.2022.36-1-gca99a1d3
.git/config present: true
thirdparty/tcc status: thirdparty-linux-amd64 827f7452

```

**What did you do?**
`v -g -o vdbg cmd/v && vdbg problem1.v`
```v
type Expr = fn ([]Expr) Expr | int

fn f(exprs []Expr) Expr {
	return 0
}

fn main() {
	mut m := map[string]Expr{}
	m['f'] = f
}

```

**What did you expect to see?**

no build error


**What did you see instead?**
```
==================
/tmp/v_1000/problem1.17519511473305661447.tmp.c:2026: error: ',' expected (got "x")
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang

```

--- a/vlib/v/checker/checker.v
+++ b/vlib/v/checker/checker.v
@@ -519,6 +519,16 @@ pub fn (mut c Checker) sum_type_decl(node ast.SumTypeDecl) {
 					}
 				}
 			}
+		} else if sym.info is ast.FnType {
+			func := (sym.info as ast.FnType).func
+			if c.table.sym(func.return_type).name.ends_with('.$node.name') {
+				c.error('sum type `$node.name` cannot be defined recursively', variant.pos)
+			}
+			for param in func.params {
+				if c.table.sym(param.typ).name.ends_with('.$node.name') {
+					c.error('sum type `$node.name` cannot be defined recursively', variant.pos)
+				}
+			}
 		}
 
 		if sym.name.trim_string_left(sym.mod + '.') == node.name {

--- a/vlib/v/checker/tests/sumtype_define_recursively.out
+++ b/vlib/v/checker/tests/sumtype_define_recursively.out
@@ -1,0 +1,5 @@
+vlib/v/checker/tests/sumtype_define_recursively.vv:1:13: error: sum type `Expr` cannot be defined recursively
+    1 | type Expr = fn ([]Expr) Expr | int
+      |             ~~~~~~~~~~~~~~~~
+    2 | 
+    3 | fn f(exprs []Expr) Expr {

--- a/vlib/v/checker/tests/sumtype_define_recursively.vv
+++ b/vlib/v/checker/tests/sumtype_define_recursively.vv
@@ -1,0 +1,10 @@
+type Expr = fn ([]Expr) Expr | int
+
+fn f(exprs []Expr) Expr {
+	return 0
+}
+
+fn main() {
+	mut m := map[string]Expr{}
+	m['f'] = f
+}

